### PR TITLE
[c++] Do not add dimensions for dense array queries if not requested

### LINKED ISSUE #2438: [c++] `DenseNDArray` read is 25X slower with tiledb-soma vs tiledb
I have a very large dense array, created as a SOMA DenseNDArray.  Reading the object with `tiledbsoma` is over 25X slower than reading the same slice using `tiledb`.  I would expect them to be more or less identical.

Sample code and example run which shows that a small slice (10000 x ~50000) takes ~12 seconds with tiledbsoma, but only 470ms with tiledb.

I have the same data mastered as a sparse SOMA ndarray, and it takes ~900ms. (not relevant to this bug report)

```

In [42]: import tiledbsoma as soma
    ...: import tiledb
    ...: 
    ...: PATH = "tmp/mouse_X_raw_dense"
    ...: RNG = (1_000_000, 1_010_000)
    ...: 
    ...: 
    ...: print("TileDB-SOMA Dense read")
    ...: somaDX = soma.open(PATH)
    ...: %timeit somaDX.read(coords=(slice(RNG[0],RNG[1]),))
    ...: somaDX.close()
    ...: 
    ...: print("TileDB-Py Dense read (same data, same array)")
    ...: tdbDX = tiledb.open(PATH)
    ...: %timeit tdbDX[RNG[0]:RNG[1]]['soma_data']
    ...: tdbDX.close()
    ...: 
    ...: print("TileDB-SOMA sparse read (same data, diff array)")
    ...: somaSX = soma.open("tmp/mouse_X_raw")  # sparse
    ...: %timeit somaSX.read(coords=(slice(RNG[0],RNG[1]),)).tables().concat()
    ...: somaSX.close()
    ...: 
TileDB-SOMA Dense read
12.6 s ± 57.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
TileDB-Py Dense read (same data, same array)
470 ms ± 1.99 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
TileDB-SOMA sparse read (same data, diff array)
926 ms ± 35.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [43]: soma.show_package_versions()
tiledbsoma.__version__        1.9.5
TileDB-Py tiledb.version()    (0, 27, 1)
TileDB core version           2.21.1
libtiledbsoma version()       libtiledb=2.21.1
python version                3.11.8.final.0
OS version                    Linux 6.5.0-1016-aws
```

Array schema:
```
In [45]: tiledb.open(PATH).schema.dump()
- Array type: dense
- Cell order: row-major
- Tile order: row-major
- Capacity: 100000
- Allows duplicates: false
- Coordinates filters: 1
  > ZSTD: COMPRESSION_LEVEL=-1
- Offsets filters: 3
  > DOUBLE_DELTA: COMPRESSION_LEVEL=-1, REINTERPRET_DATATYPE=ANY
  > BitWidthReduction: BIT_WIDTH_MAX_WINDOW=256
  > ZSTD: COMPRESSION_LEVEL=-1
- Validity filters: 1
  > RLE: COMPRESSION_LEVEL=-1

### Dimension ###
- Name: soma_dim_0
- Type: INT64
- Cell val num: 1
- Domain: [0, 41187080]
- Tile extent: 128
- Filters: 1
  > ZSTD: COMPRESSION_LEVEL=3

### Dimension ###
- Name: soma_dim_1
- Type: INT64
- Cell val num: 1
- Domain: [0, 52370]
- Tile extent: 52371
- Filters: 1
  > ZSTD: COMPRESSION_LEVEL=3

### Attribute ###
- Name: soma_data
- Type: FLOAT32
- Nullable: false
- Cell val num: 1
- Filters: 2
  > ByteShuffle
  > ZSTD: COMPRESSION_LEVEL=15
- Fill value: nan
None 
```

The array is rather large, but I can provide upon request.

### PR DESCRIPTION
**Pull request by johnkerl**
**Issue and/or context:** Resolves #2438

**Changes:** This is @nguyenv 's branch and her work for C++ and Python; I'm simply PR'ing it. For R, it's joint work with @mojaveazure.

**Notes for Reviewer:** Performance successfully evaluated in a sandbox environment using the repro on #2438.

### CODE REVIEW

**Comment by johnkerl** [comment]
Manually I found this solves the performance problem for R as well. Adapting the Python script from #2438 to R I see:

```
library(tiledbsoma)

DPATH <- "s3://tiledb-johnkerl/czi-bruce-dense/mouse_X_normalized_dense"
ctx <- tiledbsoma::SOMATileDBContext$new(c(soma.init_buffer_bytes="4000000000"))
dnda <- SOMAOpen(DPATH, tiledbsoma_ctx=ctx)

t1 <- Sys.time()
x <- dnda$read_dense_matrix(coords=list( 1000000:1005000, 1:52370 ))
t2 <- Sys.time()

cat("SECONDS", t2-t1, "\n")
```

producing

```
Warning message:
In matrix(as.numeric(tbl$GetColumnByName("soma_data")), nrow = length(unique(as.numeric(tbl$GetColumnByName("soma_dim_0")))),  :
  non-empty data for zero-extent matrix
SECONDS 5.150843
```

-- the execution time is now on par with expectations.

To do at this point:

* Resolve the above warning
* Resolve the unit-test fails (appearing in CI here)

**Review by nguyenv — APPROVED** [review]
Glad to hear this actually worked! I'm not sure what I missed either...

**Comment by johnkerl** [comment]
Debug note:

* In R: `options(warn = 2)` and `source("repro.r")` and `traceback()`
* Points to here: https://github.com/single-cell-data/TileDB-SOMA/blob/1.9.5/apis/r/R/SOMADenseNDArray.R#L83-L88

**Comment by johnkerl** [comment]
Returning to draft status until I address needed changes

**Comment by johnkerl** [comment]
Status: remaining R CI fails are due to this still needing to be adapted:
https://github.com/single-cell-data/TileDB-SOMA/blob/44b8c28d5f4315cab58e6190c6bb0e0fc338ca3f/apis/r/R/SOMAExperimentAxisQuery.R#L426-L447

**Comment by johnkerl** [comment]
Merged in #2500, and merged atop latest `main`

File: apis/r/tools/r-ci.sh [hunk 1]
```diff
@@ -372,15 +372,24 @@ DumpLogsByExtension() {
...
     extension=$1
     shift
-    package=$(find . -maxdepth 1 -name "*.Rcheck" -type d)
-    if [[ ${#package[@]} -ne 1 ]]; then
-        echo "Could not find package Rcheck directory, skipping log dump."
+
+    # Note: this script may be run from the apis/r subdirectory or from the
+    # package base directory. This runs in CI from a clean image so in the
+    # (hypothetical) case of multiple log files, we err on the side of providing
+    # the operator with more inforation rather than less.
```
Comment (eddelbuettel) [inline apis/r/tools/r-ci.sh:379]: There is a way to echo with a colon and some other marker to create the separate little sections in a scrollable log. I thought I used that somewhere but I can't locate it right now, and can't find it in the docs for GHA. 😿

Comment (johnkerl) [reply apis/r/tools/r-ci.sh:379]: Sounds like a nice to have!

Comment (johnkerl) [reply apis/r/tools/r-ci.sh:379]: Happy to take this mod on a follow-on PR

**Comment by johnkerl** [comment]
@nguyenv and @mojaveazure thanks for your patience and help, respectively!!! This is ready for review

File: apis/r/tools/r-ci.sh (hunk 1 above)
Comment (eddelbuettel) [reply apis/r/tools/r-ci.sh:379]: This seems to be it 
https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#grouping-log-lines

Comment (johnkerl) [reply apis/r/tools/r-ci.sh:379]: Thanks!

https://github.com/single-cell-data/TileDB-SOMA/pull/2453/commits/1b22704a67041de98bacfbb0ca75f63e59925600

**Comment by johnkerl** [comment]
@mojaveazure -- @nguyenv had approved this two weeks ago -- that was before the R mods -- do either of you mind to re-stamp?

**Review by nguyenv — APPROVED** [review]
I can't speak on the R code but will go ahead and approve since the C++ works and CI is all passing.

**Review by mojaveazure — APPROVED** [review]
Sorry, just saw this. LGTM, thanks for taking care of this!

**Comment by eddelbuettel** [comment]
@nguyenv That works :)   I had glanced at the R code a few times in the last few days and it does The Right Thing (TM) by filling in domain data that dense does not retrieve and does it in a logical manner (of then generating what sparse would have come back with) before using joint code.  I have not had time to dig deeper or at the C++/Python bits.

**Pull request merged**